tracing-subscriber/src/fmt/fmt_layer.rs

```diff
@@ -458,6 +521,83 @@ where
         }
     }
 
+    fn on_enter(&self, id: &Id, ctx: Context<'_, S>) {
+        let span = ctx.span(id).expect("Span not found, this is a bug");
+        let mut extensions = span.extensions_mut();
+        if let Some(timings) = extensions.get_mut::<Timings>() {
+            let now = Instant::now();
+            timings.idle += (now - timings.last).as_nanos() as u64;
+            timings.last = now;
+        }
+
+        if self.fmt_span.trace_enter() {
+            event_from_span!(event = id, span, message = "enter");
+            drop(extensions);
+            drop(span);
+            self.on_event(&event, ctx);
+        }
+    }
+
+    fn on_exit(&self, id: &Id, ctx: Context<'_, S>) {
+        let span = ctx.span(id).expect("Span not found, this is a bug");
+        let mut extensions = span.extensions_mut();
+        if let Some(timings) = extensions.get_mut::<Timings>() {
+            let now = Instant::now();
+            timings.busy += (now - timings.last).as_nanos() as u64;
+            timings.last = now;
+        }
+
+        if self.fmt_span.trace_exit() {
+            event_from_span!(event = id, span, message = "exit");
+            drop(extensions);
+            drop(span);
+            self.on_event(&event, ctx);
+        }
+    }
+
+    fn on_close(&self, id: Id, ctx: Context<'_, S>) {
+        let span = ctx.span(&id).expect("Span not found, this is a bug");
+        let extensions = span.extensions();
+        if self.fmt_span.trace_close() {
+            if let Some(timing) = extensions.get::<Timings>() {
+                let Timings {
+                    busy,
+                    mut idle,
+                    last,
+                } = *timing;
+                idle += (Instant::now() - last).as_nanos() as u64;
+
+                let mut s_idle = BufVec::new();
+                self.fmt_span.fmt_timing(idle, &mut s_idle);
+                // safe because we control the writers and they write strings
+                let s_idle = unsafe { from_utf8_unchecked(s_idle.as_ref()) };
+                let t_idle = field::display(s_idle);
+
+                let mut s_busy = BufVec::new();
+                self.fmt_span.fmt_timing(busy, &mut s_busy);
+                // safe because we control the writers and they write strings
+                let s_busy = unsafe { from_utf8_unchecked(s_busy.as_ref()) };
+                let t_busy = field::display(s_busy);
```

Comment:
The above should not allocate unless `smallvec` is deselected — but I just realize what you meant, will try to rewrite as `Display` instance.